src/function/gds/K_Core_Decomposition.cpp

```diff
@@ -0,0 +1,324 @@
+                }
+            }
+        }
+        uint64_t current = curSmallestDegree.load(std::memory_order_relaxed);
+        curSmallestDegree.compare_exchange_strong(current, curSmallest, std::memory_order_relaxed);
+    }
+
+    void setAllActive() {
+        for (const auto& [tableID, curNumNodes] : numNodesMap) {
+            curDenseFrontier->pinTableID(tableID);
+            for (uint64_t offset = 0; offset < curNumNodes; ++offset) {
+                common::nodeID_t nodeID;
+                nodeID.tableID = tableID;
+                nodeID.offset = offset;
+                curDenseFrontier->setActive(nodeID);
+            }
+        }
+    }
+
+    uint64_t getSmallestDegree() { return curSmallestDegree.load(std::memory_order_relaxed); }
+
+    void setUpdateFlag() { updateDegreeFlag = true; }
+
+    bool isNodeActive(common::nodeID_t nodeID) const {
+        return curDenseFrontier->isActive(nodeID.offset);
+    }
+
+    uint64_t getVertexValue(common::offset_t offset) {
+        return vertexValues[offset].load(std::memory_order_relaxed);
+    }
+
+private:
+    bool updateDegreeFlag = false;
+    std::atomic<uint64_t> curSmallestDegree{UINT64_MAX};
+    common::table_id_map_t<common::offset_t> numNodesMap;
+    std::atomic<uint64_t>* vertexValues = nullptr;
+    ObjectArraysMap<std::atomic<uint64_t>> curVertexValues;
+};
+
+struct KCoreInitEdgeCompute : public EdgeCompute {
+    KCoreFrontierPair* frontierPair;
+
+    explicit KCoreInitEdgeCompute(KCoreFrontierPair* frontierPair) : frontierPair{frontierPair} {}
+
+    std::vector<common::nodeID_t> edgeCompute(common::nodeID_t boundNodeID,
+        graph::NbrScanState::Chunk& chunk, bool) override {
+        std::vector<common::nodeID_t> result;
+        uint64_t nbrAmount = 0;
+        chunk.forEach([&](auto /* nbrNodeID */, auto) { nbrAmount++; });
+        result.push_back(boundNodeID);
+        frontierPair->addToVertexDegree(boundNodeID, nbrAmount);
+        return result;
+    }
+
+    std::unique_ptr<EdgeCompute> copy() override {
+        return std::make_unique<KCoreInitEdgeCompute>(frontierPair);
+    }
+};
+
+struct KCoreEdgeCompute : public EdgeCompute {
+    KCoreFrontierPair* frontierPair;
+
+    explicit KCoreEdgeCompute(KCoreFrontierPair* frontierPair) : frontierPair{frontierPair} {}
+
+    std::vector<common::nodeID_t> edgeCompute(common::nodeID_t boundNodeID,
+        graph::NbrScanState::Chunk& chunk, bool) override {
+        std::vector<common::nodeID_t> result;
+        uint64_t vertexDegree = frontierPair->getVertexValue(boundNodeID.offset);
+        uint64_t smallestDegree = frontierPair->getSmallestDegree();
+        if (frontierPair->isNodeActive(boundNodeID) && (vertexDegree == smallestDegree)) {
+            chunk.forEach([&](auto nbrNodeID, auto) {
+                if (frontierPair->isNodeActive(nbrNodeID)) {
+                    frontierPair->removeFromVertex(nbrNodeID);
+                    result.push_back(nbrNodeID);
```

Comment:
You are making every node whose degree has been updated (fetch_sub(1)) as active nodes in next frontier. Though I think you only need those whose degree equals to new smallest degree as active nodes right? 

In K-core algorithm, as I understand, frontier size should keep decreasing because as K increases, more nodes are removed. Your implementation doesn't seem to have such property.

Also, I believe there is racing condition here. You are reading vertexDegree in l160 and updating vertexDegree in l165. There is no guarantee that the value in l160 has not been updated by the time you use it.

Reply:
Let's discuss in person. I don't understand the difference between the output of Ligra & Neo4j.

Let me give a more concrete example of racing condition.

Suppose node A has degree 5 and node B has degree 4 in current iteration. There is one edge between A&B. And we are looking for smallestDegree == 4. 

Logically, you want to only consider B in the current iteration (core). However, in a racing condition, you might first visit B, compute it, removeFromVertex for its neighbours, i.e. A and reduce A's degree to 4. Then u see A, which now has degree == 4 and you will consider A in current iteration as well.

Is above a legit racing condition? If so, add a test case to reveal it and then fix it.